onnxruntime/core/providers/cuda/tensor/scatter_elements.cc

```diff
@@ -1,176 +1,117 @@
-  if (input_dims.size() != indices_dims.size()) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices must have the same rank as Input. Indices rank=",
-                           indices_dims.size(), ". Input rank=", input_dims.size());
-  }
-
-  for (size_t i = 0; i < input_dims.size(); ++i) {
-    // For all axes except the axis of interest, make sure that the corresponding 'indices' shape
-    // value is within bounds of the corresponding 'data' shape.
-    if (static_cast<int64_t>(i) != axis_ && input_dims[i] < indices_dims[i]) {
-      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices dim=", indices_dims[i], " at pos=", i,
-                             " is greater than input dim=", input_dims[i]);
-    }
+  const auto& indices_shape = indices_tensor->Shape();
+  const int64_t indices_size = indices_shape.Size();
+  if (indices_shape != updates_tensor->Shape()) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Indices and updates must have the same shape.");
   }
 
-  int rank = (int)input_dims.size();
-  auto* output_tensor = context->Output(0, input_data_shape);
-
-  TArray<int64_t> buffer_input_dims(input_dims);
-  TensorPitches input_strides(input_dims);
-  TArray<int64_t> buffer_input_strides(input_strides);
-
-  TArray<int64_t> buffer_indices_dims(indices_dims);
-  TArray<fast_divmod> fdm_indices_strides(rank);
-  TensorPitches indices_strides(indices_dims);
-  for (auto i = 0; i < rank; i++) {
-    fdm_indices_strides[i] = fast_divmod(static_cast<int>(indices_strides[i]));
+  // Validate input shapes and ranks (invoke the static method in the CPU GatherElements kernel that hosts the shared
+  // checks)
+  ORT_RETURN_IF_ERROR(onnxruntime::GatherElements::ValidateInputShapes(input_shape, indices_shape, axis));
+
+  auto* output_tensor = context->Output(0, input_shape);
+  if (input_size == 0) return Status::OK();
+
+  TensorShapeVector input_shape_vec = input_shape.AsShapeVector();
+  TensorShapeVector indices_shape_vec = indices_shape.AsShapeVector();
+  int64_t new_axis, new_rank, input_stride_along_axis;
+  TArray<int64_t> masked_input_strides;
+  TArray<fast_divmod> indices_fdms;
+  CoalesceDimensions(input_shape_vec, indices_shape_vec, axis, new_axis, new_rank, input_stride_along_axis,
+                     masked_input_strides, indices_fdms);
+
+  // Use element size instead of concrete types so we can specialize less template functions to reduce binary size.
+  int dtype = GetElementType(input_tensor->DataType()->Size());
+  if (dtype == ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED) {
+    ORT_THROW("Unsupported element size by the ScatterElements CUDA kernel");
   }
 
-  utils::MLTypeCallDispatcher<float, MLFloat16, int16_t, int8_t, int32_t,
-                              int64_t, uint8_t, uint16_t, uint32_t, uint64_t, double, bool>
-      t_disp(data_tensor->GetElementType());
+  utils::MLTypeCallDispatcher<int8_t, MLFloat16, float, double> t_disp(dtype);
```

Comment:
why we don't use "t_disp(data_tensor->GetElementType());" directly, or "t_disp(input_tensor->GetElementType());"

Reply:
This is because we just specializes the template functions for 4 data types (for less specialization code in kernel file and smaller binary size), as mentioned in above comment. If we use GetElementType, then we need to specialize the template functions for all these types.